[Luo Zhijie] iP

Jay-luozhijie opened the pull request:
# Duke chat bot :robot:

>Creativity is intelligence having FUN - Albert Einstein

This is the chat bot that lets you manage your tasks easily!([download it from here](https://github.com/LuoZhijie-tom/ip.git))

### Commands:
* __todo__ *description* : Add a ToDo task.
* __deadline__ *description* __/by__ *deadline* : Add a Deadline task.
* __event__ *description* __/at__ *event time* : Add an Event.
* __done__ *taskIndex*: Marks the task at *taskIndex* as done.
* __list__ : See all existing tasks.
* __find__ *keyword* : Find all tasks with *keyword*.
* __delete__ *taskIndex* : Delete task at *taskIndex*.

### Current features:
1. add delete and mark three types of tasks which are ToDo, Deadline, and Event.
2. mark the task as done.
3. list all existing tasks.
4. search for tasks based on keyword.

### Features:
- [X] Managing tasks(add, delete, mark as done, list, find)
- [ ] GUI

### Sample codes(`run()` method):
```java
public void run() {
        try {
            tasks = storage.convertFileToTaskList();
        } catch (Exception e) {
            System.out.println(e.getMessage());
            return;
        }
        ui.greet();
        boolean isExit = false;
        Parser parser = new Parser(" ");
        while (!isExit) {
            try {
                String fullCommand = ui.readCommand();
                Command c = parser.parse(fullCommand);
                c.execute(tasks, ui, storage);
                isExit = c.isExit();
            } catch (DukeException e) {
                ui.showError(e.getMessage());
            }
        }
    }
```

Commit:
Add Gradle support

Jay-luozhijie committed:
added level-1: Greet,Echo,Exit

Jay-luozhijie committed:
added Level-2: add and list

Jay-luozhijie committed:
added Level-3: mark as done and class Task

Jay-luozhijie committed:
get the command from String.split(" ")

Jay-luozhijie committed:
add toString() in Task

Jay-luozhijie committed:
level-4 basic added

Jay-luozhijie committed:
level-4 polimophism added

Jay-luozhijie committed:
added exception handling

Jay-luozhijie committed:
A-TextUiTesting

Jay-luozhijie committed:
Merge branch 'master' into HEAD

Jay-luozhijie committed:
Level-6 added

Jay-luozhijie committed:
add level 7

Jay-luozhijie committed:
add Level-8

Jay-luozhijie committed:
Merge branch 'branch-Level-8' to master

Jay-luozhijie committed:
add more OOP

Jay-luozhijie committed:
add package duke

Jay-luozhijie committed:
add JUnit test

Jay-luozhijie committed:
add Jar

Jay-luozhijie committed:
add JavaDoc

Jay-luozhijie committed:
reformat the codes

Jay-luozhijie committed:
Reformat the code

Jay-luozhijie committed:
add find command

Jay-luozhijie committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/AddCommand.java
#	src/main/java/duke/Duke.java

Jay-luozhijie committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java

Jay-luozhijie committed:
handle file not found case

Jay-luozhijie committed:
add hour and minute

Jay-luozhijie committed:
add localDateTime with hour and minute

Jay-luozhijie committed:
Merge branch 'add-gradle-support' of https://github.com/LuoZhijie-tom/ip into A-Gradle

Jay-luozhijie committed:
add test

Jay-luozhijie committed:
add gradle

Jay-luozhijie committed:
add gradle

Jay-luozhijie committed:
add gradle run and test

Jay-luozhijie committed:
update dukeData

uyencfi commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,22 @@
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage);
+
+    /**
+     * indicates whether the command is to exit duke bot.
```
Hey the i in "indicates" should be capitalized 😃

uyencfi commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,96 @@
+     */
+    public Command parse(String input) throws DukeException {
+        String words[] = input.split(this.splitRegex);
+        if (words[0].equals("list")) {
```
I think this chain of `if-else` can be converted to `switch-case` statements to reduce the nesting level!

uyencfi commented on src/main/java/duke/MarkAsDoneCommand.java:
```diff
@@ -0,0 +1,37 @@
+        if (obj instanceof MarkAsDoneCommand) {
+            @SuppressWarnings("have checked obj is MarkAsDoneCommand, can safely parse")
+            MarkAsDoneCommand temp = (MarkAsDoneCommand) obj;
+            return temp.doneListIndex == this.doneListIndex
+                    ? true
+                    : false;
```
Can these lines be shortened as `return temp.doneListIndex == this.doneListIndex`?

uyencfi commented on src/main/java/duke/ListCommand.java:
```diff
@@ -0,0 +1,25 @@
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj instanceof ListCommand) {
+            return true;
+        } else {
+            return false;
+        }
```
Perhaps `return (obj instanceof ListCommand)` here for readability!

uyencfi commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,76 @@
+/**
+ * The task object.
+ */
+public class Task {
```
I would say `Task` can be made an abstract class to avoid instantiation of a Task that is not a ToDo, a Deadline or an Event.

uyencfi commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,69 @@
+ * ArrayList of tasks.
+ */
+public class TaskList {
+    ArrayList<Task> tasks;
```
Probably not an issue for iP but perhaps use `List<Task>` as the attribute type here to allow flexibility should you decide to switch the implementation to e.g. a `LinkedList` in the future?

uyencfi reviewed (commented):
LGTM! Just some white space issues. The code already conforms to the coding standard well so I had to comment on some non-coding standard stuff 😅

Jay-luozhijie committed:
Update README.md

Jay-luozhijie committed:
Merge branch 'master' of https://github.com/LuoZhijie-tom/ip

Jay-luozhijie committed:
Merge branch 'A-Gradle'

# Conflicts:
#	gradle/wrapper/gradle-wrapper.jar
#	gradle/wrapper/gradle-wrapper.properties
#	gradlew
#	gradlew.bat
#	src/main/java/duke/Event.java
#	src/test/java/duke/ParserTest.java
#	src/test/java/duke/StorageTest.java
#	src/test/java/test/duke/ParserTest.class
#	src/test/java/test/duke/StorageTest.class

Jay-luozhijie committed:
learn GUI

Jay-luozhijie committed:
add GUI

Jay-luozhijie committed:
Merge tag 'Level-10'

Jay-luozhijie committed:
change build.gradle duke.Duke to duke.launcher

Jay-luozhijie committed:
Add assertions

I make some assumptions about the input string format.

But users may not enter the input as expected which might cause
error while parsing.

Let's add assertions before parsing the string into command.

Adding assertions ensures that the input format matches with the
assumption so that the code will work smoothly.

Jay-luozhijie committed:
Parser class: add more parsing methods

The parse() method is too long.

This make the code diffcult to read.

Let's create more parsing method to handle different commands.

The parse() method becomes shorter and easier to use

Jay-luozhijie committed:
Parser class : delete magic numbers

There are some magic numbers in the parse(), parseDeadline(), and parseEvent() method.

The magic numbers make code harder to understand.

Let's create some variable that represent the magic numbers.

This makes code easier to understand

Jay-luozhijie committed:
modify fxml

Jay-luozhijie committed:
rename some variable

Jay-luozhijie committed:
Merge pull request #1 from LuoZhijie-tom/A-CodeQuality

A code quality

Jay-luozhijie committed:
Add duke greeting

Jay-luozhijie committed:
Add assertion

The code assume that task description won't contain sign "&&" thus "&&" is used to split the commands.

But if there is "&&" inside task description, the parser won't work properly.

Let's add an assertion while writing tasks into the file that the task description does not contain "&&".

After that, duke can read from the txt file safely.

Jay-luozhijie committed:
Merge pull request #2 from LuoZhijie-tom/A-CodeQuality

Add duke greeting

Jay-luozhijie committed:
Merge branch 'master' into branch-A-Assertions

Jay-luozhijie committed:
Merge branch 'branch-A-Assertions'

Jay-luozhijie committed:
learning gui

Jay-luozhijie committed:
Add todo read from txt reminder

Jay-luozhijie committed:
Add reminder

Jay-luozhijie committed:
change GUI

Jay-luozhijie committed:
Add Ui.png

Jay-luozhijie committed:
Update README.md

Jay-luozhijie committed:
Update README.md

Jay-luozhijie committed:
Update README.md

Jay-luozhijie committed:
Update README.md

Jay-luozhijie committed:
Add User guide

Jay-luozhijie committed:
refine code

Jay-luozhijie committed:
Refine code style

Jay-luozhijie committed:
Add more test cases

Jay-luozhijie committed:
Change code styles

Jay-luozhijie committed:
Delete some tasks.

Jay-luozhijie committed:
Update README.md

Jay-luozhijie committed:
Update README.md

Jay-luozhijie committed:
Change Task to be abstract

Jay-luozhijie committed:
Modify parser

Jay-luozhijie committed:
Add empty file cases when command "list"

Jay-luozhijie committed:
Fix error in unit test

Jay-luozhijie committed:
Delete redundant isExit() method in Command

Jay-luozhijie committed:
Update README.md

Jay-luozhijie committed:
Split parseDeadline() and parseEvent() method to more methods

Jay-luozhijie committed:
Update README.md

Jay-luozhijie committed:
Update README.md
